### PR TITLE
Support UDP flows and Fill L3/L4 checksums

### PR DESCRIPTION
**Pull request by shinae-woo**
With this PR, `Flowgen` supports UDP flows along with existing TCP flows, and also fill L3/L4 checksums appropriately.

### CODE REVIEW

File: core/modules/flowgen.cc [hunk 1]
```diff
@@ -405,26 +435,67 @@ void FlowGen::DeInit() {
...
 
   Ipv4 *ipheader = reinterpret_cast<Ipv4 *>(p + sizeof(Ethernet));
-  Tcp *tcpheader = reinterpret_cast<Tcp *>(p + sizeof(Ethernet) + sizeof(Ipv4));
-
   ip_src_base_ = ipheader->src.value();
   ip_dst_base_ = ipheader->dst.value();
-  port_src_base_ = tcpheader->src_port.value();
-  port_dst_base_ = tcpheader->dst_port.value();
+
+  if (l4_proto_ == Ipv4::Proto::kUdp) {
```
Comment (sangjinhan) [inline core/modules/flowgen.cc]: Most existing code of other modules (e.g., ipswap or nat) does not differentiate TCP and UDP when only L4 ports are concerned, to reduce code redundancy. UDP and TCP share the same layout for src/dst ports for this purpose.

```
Udp *l4_header = reinterpret_cast<Udp *>(p + sizeof(Ethernet) + sizeof(Ipv4));
port_src_base_ = l4_header->src_port.value();
port_dst_base_ = l4_header->dst_port.value();
```
should be enough for both UDP and TCP. You may also want to declare `using L4 = Udp;` somewhere to feel less guilty...

File: core/modules/flowgen.h [hunk 2]
```diff
@@ -114,7 +114,8 @@ class FlowGen final : public Module {
   void PopulateInitialFlows();
 
   CommandResponse UpdateBaseAddresses();
-  bess::Packet *FillPacket(struct flow *f);
+  bess::Packet *FillUDPPacket(struct flow *f);
```
Comment (sangjinhan) [inline core/modules/flowgen.h]: Given that `FillUDPPacket` (btw, the style guide suggests `FillUdpPacket`, which I personally feel uglier) and `FillTcpPacket` share the considerable amount of logic, Could you merge them into `FillPacket` again but add a conditional for L4-specific routine inside the function?

File: core/modules/flowgen.cc [hunk 3]
```diff
@@ -381,8 +411,8 @@ CommandResponse FlowGen::Init(const bess::pb::FlowGenArg &arg) {
     return CommandFailure(ENOMEM, "task creation failed");
   }
 
-  templ_ = new char[MAX_TEMPLATE_SIZE];
-  if (templ_ == nullptr) {
+  tmpl_ = new char[MAX_TEMPLATE_SIZE];
```
Comment (sangjinhan) [inline core/modules/flowgen.cc]: Hm... although it's an existing issue and not part of this PR, I am not sure why `tmpl_` is dynamically allocated here. Could you please make it `unsigned char tmpl_[MAX_TEMPLATE_SIZE];` member of the module class?

File: core/modules/flowgen.cc [hunk 4]
```diff
@@ -282,8 +297,23 @@ CommandResponse FlowGen::CommandUpdate(const bess::pb::FlowGenArg &arg) {
...
-    memset(templ_, 0, MAX_TEMPLATE_SIZE);
-    bess::utils::Copy(templ_, arg.template_().c_str(), template_size_);
+    const char *tmpl = arg.template_().c_str();
+    const Ethernet *eth = reinterpret_cast<const Ethernet *>(tmpl);
+    if (eth->ether_type != be16_t(Ethernet::Type::kIpv4)) {
+      return CommandFailure(EINVAL, "'template' is not IPv4");
+    }
+
+    const Ipv4 *ip = reinterpret_cast<const Ipv4 *>(eth + 1);
+    if (ip->protocol != Ipv4::Proto::kUdp and ip->protocol != Ipv4::Proto::kTcp) {
```
Comment (sangjinhan) [inline core/modules/flowgen.cc]: `and` is here too

File: core/modules/flowgen.cc [hunk 5]
```diff
@@ -282,8 +297,23 @@ CommandResponse FlowGen::CommandUpdate(const bess::pb::FlowGenArg &arg) {
...
+      return CommandFailure(EINVAL, "'template' is not UDP or TCP");
+    }
+
+    if (l4_proto_ != ip->protocol) {
```
Comment (sangjinhan) [inline core/modules/flowgen.cc]: Q: is there any reason to not change the L4 protocol of template?

File: core/modules/flowgen.cc [hunk 6]
```diff
@@ -405,26 +435,67 @@ void FlowGen::DeInit() {
...
   Ipv4 *ipheader = reinterpret_cast<Ipv4 *>(p + sizeof(Ethernet));
-  Tcp *tcpheader = reinterpret_cast<Tcp *>(p + sizeof(Ethernet) + sizeof(Ipv4));
-
   ip_src_base_ = ipheader->src.value();
   ip_dst_base_ = ipheader->dst.value();
-  port_src_base_ = tcpheader->src_port.value();
-  port_dst_base_ = tcpheader->dst_port.value();
+
+  if (l4_proto_ == Ipv4::Proto::kUdp) {
+    Udp *udpheader = reinterpret_cast<Udp *>(p + sizeof(Ethernet) + sizeof(Ipv4));
```
Comment (sangjinhan) [inline core/modules/flowgen.cc]: existing issue: IP header with options

File: core/modules/flowgen.h [hunk 7]
```diff
@@ -129,8 +130,9 @@ class FlowGen final : public Module {
   // Priority queue of future events
   EventQueue events_;
 
-  char *templ_;
+  char *tmpl_;
   int template_size_;
+  uint16_t l4_proto_;
```
Comment (sangjinhan) [inline core/modules/flowgen.h]: this should be `uint8_t`? (to match the type of `ip->protocol`)

File: core/modules/flowgen.cc (hunk 1 above)
Comment (shinae-woo) [reply core/modules/flowgen.cc]: Sad :(

File: core/modules/flowgen.cc (hunk 5 above)
Comment (shinae-woo) [reply core/modules/flowgen.cc]: Just because current implementation doesn't differentiate new flows (with changed protocol) from the existing flows (old protocol)

**Comment by sangjinhan** [comment]
Merged, thanks!
n90tpt6kwfqhzbmvnfzmm0xlo1_500
